WIP: Unwrapping UnionAll types for datatype definition in REPL help

m-j-w opened the pull request:
This addresses  #21029. The proposed solution has served me well today, so maybe it's worth a PR.
I added tests. Still, I'm unsure whether this solves the issue completely.

Commit:
Add specialization for UnionAll types to Base.Docs.summarize(...)

Commit:
Added tests for unwrapped unionall on 'struct', 'primitive type' and 'abstract type'.

JeffBezanson commented:
This is probably basically the right thing to do, but note that the result of `unwrap` is not a valid type, so for example the `subtypes` call might not work.

m-j-w commented:
How would I solve this potential subtype printing issue? I didn't observe wrong printing of subtypes of abstract types. Neither with the same or additional type parameters.
```jl
julia> abstract type a{T<:Number,N} end
       primitive type p{T<:Number,N,A}<:a{T,N} 32 end
       struct s{T<:Number,N} <: a{T,N}
          a::NTuple{N,T}
       end

help?> a
search: any all abs Any ANY atan asin asec any! all! airy acsc acot acos abs2 atanh atand atan2 asinh asind asech asecd ascii angle airyx acsch acscd acoth

  No documentation found.

  Summary:

  abstract type a{T<:Number,N} <: Any

  Subtypes:

  p{T<:Number,N,A} where A
  s{T<:Number,N}

  Note, a is represented by a wrapping type UnionAll that describes the bounds of its type parameters.
```

JeffBezanson commented:
Yes, it might happen to work in a number of cases. See the `subtypes` function itself for a good example --- it keeps both a wrapped and unwrapped version of a type, and uses the unwrapped version when it needs to look at the structure and the wrapped version when it needs to call `typeintersect`. That's generally a good pattern to use.

m-j-w commented:
I see. I'll try to work that out.

Commit:
Second attempt to summarize `UnionAll` and `Union`.

Printing of subtypes having different upper and lower type parameter bounds
seems now to work.  Used the following type below for testing.

However, this doesn't seem to take yet the issue mentioned by Jeff into account.

This also contains an initial attempt to summarize `Union`s; unfortunately
`UnionAll`s of `Union`s are a bit tricky... (`Union` are not printed in Julia 0.5 either.)

```jl
abstract type at0{T<:Number,N} end
abstract type at1{T>:Integer,N} <:at0{T,N} end

primitive type pt2{T<:Number,N,A>:Integer} <:at0{T,N} 32 end

struct st3{T<:Integer,N} <: at0{T,N}
   a::NTuple{N,T}
   b::Array{Int,N}
   c::Int
end

struct st4{T,N} <: at0{T,N}
   a::T
   b::NTuple{N,T}
end

struct st5{T>:Int,N} <:at1{T,N}
   c::st3{T,N}
end

mutable struct mt6{T<:Integer,N} <:at1{T,N}
   d::st5{T,N}
end

const ut7 = Union{st5, mt6}
const ut8 = Union{at1, pt2, st3, st4}
const ut9{T} = Union{at1{T}, pt2{T}, st3{T}, st4{T}}
```

m-j-w commented:
Accidental push and still WIP. (I'm aware of the unnecessary additional scopes...)

m-j-w commented:
@JeffBezanson would you mind giving 11a26c1 a quick look ?
To me it seems `UnionAll` are printed correctly. I also gave `Union` a try which isn't yet handled either, but a `UnionAll` of a `Union` is kind of tricky; and I'm not really sure I understand the overall approach taken e.g. in `show(IOContext(io, :unionall_env => x.var), x.body)`. Thanks !

JeffBezanson commented on base/docs/Docs.jl:
```diff
@@ -373,28 +373,31 @@ function summarize(io::IO, λ::Function, binding)
+function summarize(io::IO, T::Type{Union}, binding)
+    println(io, "**Types:**")
+    t = getfield(binding.mod, binding.var)
+    for ut in Core.Inference.uniontypes(Base.unwrap_unionall(t))
```
`uniontypes` also exists in Base.

JeffBezanson commented on base/docs/Docs.jl:
```diff
@@ -373,28 +373,31 @@ function summarize(io::IO, λ::Function, binding)
+    t = getfield(binding.mod, binding.var)
+    for ut in Core.Inference.uniontypes(Base.unwrap_unionall(t))
+        print(io, " - `")
+        Base.show_datatype(io, isa(ut, UnionAll) ? Base.unwrap_unionall(ut) : ut)
```
I don't really understand the purpose of this code. We already have `show` methods for types; why not just call `show`?
